python/revng/scripting/project.py

```diff
@@ -0,0 +1,93 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from typing import Any, List
+
+from revng.model import Binary, get_element_by_path
+from revng.tupletree import TypedList
+
+
+class Project(ABC):
+    def __init__(self) -> None:
+        self.model = Binary()
+        self.last_saved_model = deepcopy(Binary)
+        self.pipeline_description: dict = {}
+
+    @abstractmethod
+    def import_binary(self, input_binary: str):
+        pass
+
+    @abstractmethod
+    def get_artifact(self, artifact_name: str, targets_class: List[Any] = []) -> bytes | str:
+        pass
+
+    def get_artifacts(self, params: dict[str, List[Any]]) -> dict:
+        result = {}
+        for artifact_name, targets in params.items():
+            result[artifact_name] = self.get_artifact(artifact_name, targets)
+
+        return result
+
+    @abstractmethod
+    def analyze(self, analysis_name: str, target_class: List[Any] = []) -> Binary:
+        pass
+
+    @abstractmethod
+    def get_targets(self, target_class: List[Any], artifact_name: str) -> List[Any]:
+        pass
+
+    @abstractmethod
+    def commit(self) -> None:
+        pass
+
+    def set_model(self, model: Binary) -> None:
+        self.model = model
+        self.last_saved_model = deepcopy(self.model)
+
+        if not self.pipeline_description:
+            return
+
+        for ct in self.pipeline_description["Ranks"]:
+            model_path = ct.get("ModelPath")
+            if not model_path:
+                continue
+
+            obj = get_element_by_path(model_path, self.model)
+            if isinstance(obj, TypedList):
+                for elem in obj:
+                    elem._project = self
+            else:
+                obj._project = self
+
+    def _get_analysis_inputs(self, analysis_name: str) -> dict:
+        inputs: dict = {}
+        for step in self.pipeline_description["Steps"]:
+            for analysis in step["Analyses"]:
+                if analysis["Name"] == analysis_name:
+                    if not analysis["ContainerInputs"]:
+                        return inputs
+                    for i in analysis["ContainerInputs"]:
+                        inputs[i["Name"]] = i["AcceptableKinds"][0]
```

Comment:
Drop the `[0]`, all kinds should be returned and handled

Reply:
How should the `Kinds` be passed to the `gql` query? Currently the targets are passed like this: 
```
{'module.ll': ['0x401000:Code_x86_64:stack-pointer-promoted', '0x401010:Code_x86_64:stack-pointer-promoted', ...]}
```
Should I make two sets of inputs if more than one `Kinds`? So you would end up with: 
```
{
  'module.ll': ['0x401000:Code_x86_64:stack-pointer-promoted', '0x401010:Code_x86_64:stack-pointer-promoted', ...],
  'module.ll': ['0x401000:Code_x86_64:some-other-kind, '0x401010:Code_x86_64:some-other-kind', ...]
}
```

Reply:
No, all kinds should be under `module.ll`, you'd make another key if there was another container to show

Reply:
You mean like this:
```
{
  'module.ll': ['0x401000:Code_x86_64:stack-pointer-promoted', '0x401000:Code_x86_64:some-other-kind', ...
}
```
Or this:
```
{
  'module.ll': ['0x401000:Code_x86_64:stack-pointer-promoted:some-other-kind', ...
}
```
Or a different way?

Reply:
The first